<launch>
  <arg name="use_simulation" value="true"/>
  <arg name="template_dir" value="$(find multi_pallet_detection)/template/pallet1cell1cm.png"/>

  <node pkg="multi_pallet_detection" type="pallet_detection" name="pallet_detection" 
    output = "screen" if="$(arg use_simulation)" >
    <param name="min_x_cutoff" value="-1.2"/>
    <param name="max_x_cutoff" value="1.2"/>
    <param name="min_y_cutoff" value="0.2"/>
    <param name="max_y_cutoff" value="0.7"/>
    <param name="min_z_cutoff" value="0.2"/>
    <param name="max_z_cutoff" value="3.5"/>
    <param name="downsample_scale" value="2"/>
 
    <!-- Parameters for floor remove -->
    <param name="remove_floor" value="true"/>
    <param name="floor_distance_thresh" value="0.02"/>

    <!-- Parameters for pointcloud segementation -->
    <param name="active_segment" value="false"/> 
    <param name="opencv_show" value="true"/> 
    <param name="min_cluster_size" value="300"/>           
    <param name="max_cluster_size" value="100000"/>
    <param name="neighbor_number" value="30"/>
    <param name="curvature_thres" value="2.0"/>
    <param name="smooth_thresh" value="8.0"/>
    <param name="k_search" value="380"/>

    <!-- Normal estimation -->
    <param name="normal_estimation_method" value="2"/> 
    <param name="border_policy_ignore" value="false"/>
    <param name="publish_normal" value="true"/>
    <param name="normal_smoothing_size" value="30.0"/>
    <param name="max_depth_change_factor" value="0.1"/>
    <param name="depth_dependent_smoothing" value="true"/>

        <!-- Template matching -->
    <param name="template_directory" value="$(arg template_dir)"/>
    <param name="match_method" value="3"/>
    <param name="template_matching_show" value="true"/>
    <param name="scale_par_" value="1.0"/>
    <param name="min_scale" value="0.8"/>
    <param name="max_scale" value="2.4"/>
    <param name="manual_scale" value="false"/>
    <param name="thres" value="0.9"/> 
    <param name="change_thres" value="0.2"/> <!--1e9 -->
    <param name="scale_change_rate_" value="0.01"/>

        <!-- Kalman filter -->
    <param name="process_noise" value="0.001"/>
    <param name="measure_noise" value="0.02"/>
    <param name="pallet_gt_x" value="3.349188"/>
    <param name="pallet_gt_y" value="-0.270627"/>
    <param name="pallet_gt_yaw" value="-0.270627"/>
    <remap from = "odom" to = "RX20_16/odom"/>
  </node>

  <node pkg="multi_pallet_detection" type="pallet_detection" name="pallet_detection" 
    output = "screen" unless="$(arg use_simulation)" >
    <!-- Parameters for pointcloud cutoff -->
    <param name="min_x_cutoff" value="-1.2"/>
    <param name="max_x_cutoff" value="1.2"/>
    <param name="min_y_cutoff" value="0.0"/>
    <param name="max_y_cutoff" value="2.7"/>
    <param name="min_z_cutoff" value="0.2"/>
    <param name="max_z_cutoff" value="1.7"/>
    <param name="downsample_scale" value="2"/>

    <!-- Parameters for floor remove -->
    <param name="active_remove_floor" value="true"/>
    <param name="floor_distance_thresh" value="0.03"/>

    <!-- Parameters for pointcloud segementation -->
    <param name="active_segment" value="false"/> 
    <param name="min_cluster_size" value="1200"/>           
    <param name="max_cluster_size" value="1000000"/>
    <param name="neighbor_number" value="100"/>
    <param name="curvature_thres" value="5.0"/>
    <param name="smooth_thresh" value="10.0"/>
    <param name="k_search" value="235"/>

    <!-- Normal estimation -->
    <param name="normal_estimation_method" value="2"/> 
    <param name="border_policy_ignore" value="false"/>
    <param name="publish_normal" value="true"/>
    <param name="normal_smoothing_size" value="30.0"/>
    <param name="max_depth_change_factor" value="0.1"/>
    <param name="depth_dependent_smoothing" value="true"/>

    <!-- Template matching -->
    <param name="template_directory" value="$(arg template_dir)"/>
    <param name="match_method" value="3"/>
    <param name="opencv_show" value="true"/> 
    <param name="template_matching_show" value="true"/>
    <param name="scale_par" value="3.0"/>
    <param name="min_scale" value="2.3"/>
    <param name="max_scale" value="6.0"/>
    <param name="manual_scale" value="false"/>
    <param name="thres" value="0.9"/> <!--1e9 -->
    <param name="change_thres" value="0.12"/> 
    <param name="scale_change_rate_" value="0.01"/>

    <!-- Kalman filter -->
    <param name="process_noise" value="0.001"/>
    <param name="measure_noise" value="0.02"/>

  </node>
</launch>